src/common/indexById.ts

```diff
@@ -0,0 +1,43 @@
+interface BaseIndexableObj {
+  id: number;
+  __ignoreDuplication?: boolean;
+}
+
+const typedKeys = <T>(obj: T) => Object.keys(obj) as Array<keyof typeof obj>;
+
+interface NameIndexed<T> {
+  [name: string]: T;
+}
+
+// For ease of use we may want to be able to both access items by code names for completion/compile time checking (e.g. `ABILITIES.AURA_OF_MERCY_TALENT`) and by spell id for runtime evaluation (e.g. `ABILITIES[183415]`)
+const indexById = <Idx extends BaseIndexableObj>() => <T extends NameIndexed<Idx>>(
+  arg: T,
+): {
+  [key in keyof typeof arg | number]: Idx;
+} => {
```

Comment:
the typing of this method is what erases the specific types of keys, changing them from e.g. `Enchant` to `Spell`. This can likely be recast as:

```ts
function indexById<E extends BaseIndexableObj, T extends Record<string, BaseIndexableObj>>(originalMap: T): T & { [id: number]: E }
```

but i'll fiddle around with the specifics of this later.